codelists/views.py

```diff
@@ -105,8 +105,28 @@ def get_context_data(self, **kwargs):
 
         return context
 
-    def get_form(self, data=None, files=None):
-        return CodelistCreateForm(data, files)
+    def get_form(self, project=None, data=None, files=None):
+        """
+        Construct a Codelist ModelForm
+
+        Users arrive at this view via a URL containing a Project slug.  From
+        that we retrieve a Project (and save it to self.project) so the User
+        doesn't have to pick a Project on the form.
+
+        However Codelists both require a Project _and_ need to do validation on
+        them to check their name is unique within that Project.  ModelForms can
+        do this for us but they need to be told about the fields involved in
+        the unique_together.
+
+        Since the `project` field isn't included in the CodelistCreateForm
+        fields we pass our passed-in Project to the form using the instance
+        keyword argument so it can be used at validation time.
+        """
+        kwargs = {}
+        if project:
+            kwargs["instance"] = Codelist(project=project)
+
+        return CodelistCreateForm(data, files, **kwargs)
```

Comment:
What happens if you have `return CodelistCreateForm(data, files, instance=Codelist(project=self.project))`?

Reply:
It's identical, but I try not to rely on self outside of dispatch/the verb methods in CBVs to make testing them with View.setup() easier.

Reply:
Based on testing chat in slack, this makes use of self.project in d565d39